pkg/daemon/ceph/multus/mover.go

```diff
@@ -0,0 +1,437 @@
+	}
+
+	return controllerruntime.Result{}, nil
+}
+
+// this version of the function can be reused outside of the reconciler (i.e., the main process cleanup)
+// calling function must manage the mutex lock to make this an atomic action
+func cleanUpMovedInterface(holderPodName string, movedCache *MovedInterfaceCache) error {
+	info, err := movedCache.Get(holderPodName)
+	if err != nil {
+		return nil // no cleanup to do
+	}
+
+	// Get a handle to the host namespace
+	hostNs, err := netExec.GetCurrentNS()
+	if err != nil {
+		return errors.Wrap(err, "failed to determine host network namespace")
+	}
+
+	if err := DeleteInterface(hostNs, info.Interface); err != nil {
+		return errors.Wrapf(err, "failed to delete moved interface %q for holder pod %q", info.Interface, holderPodName)
+	}
+
+	movedCache.Remove(holderPodName)
+
+	logger.Infof("successfully cleaned up moved interface %q for holder pod %q", info.Interface, holderPodName)
+	return nil
+}
+
+func predicateAcceptOnlyHolderPods(nodeName string) predicate.Funcs {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			if !isHolderPodOnNode(e.Object, nodeName) {
+				return false
+			}
+			logger.Infof("holder pod %q was created on node %q", e.Object.GetName(), nodeName)
+			return true
+		},
+
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if !isHolderPodOnNode(e.ObjectNew, nodeName) {
+				return false
+			}
+			logger.Infof("holder pod %q was updated on node %q", e.ObjectNew.GetName(), nodeName)
+			return true
+		},
+
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			if !isHolderPodOnNode(e.Object, nodeName) {
+				return false
+			}
+			logger.Infof("holder pod %q was deleted on node %q", e.Object.GetName(), nodeName)
+			return true
+		},
+
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}
+}
+
+func isHolderPodOnNode(o client.Object, nodeName string) bool {
+	pod, ok := o.(*corev1.Pod)
+	if !ok {
+		logger.Warningf("got object %+v that is not a pod", o)
+		return false
+	}
+
+	if pod.Spec.NodeName != nodeName {
+		return false // pod is not on this mover's node
+	}
+
+	labels := pod.GetLabels()
+	if labels == nil {
+		return false // if labels are nil, it can't possibly be a holder
+	}
+	app, ok := labels[k8sutil.AppAttr]
+	if ok && app == HolderAppLabel {
+		return true // is holder pod
+	}
+
+	return false
+}
+
+func cleanupAllInterfacesWithPrefix(prefix string) {
+	// this is used for cleanup only, which we want to attempt even if there are errors
+	// we would normally use netNS.Do() here, but to minimize the chance of errors, just start
+	// in the current namespace which should be the host anyway
+
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		logger.Errorf("failed to get interfaces in host net namespace during cleanup")
+	}
+
+	for _, iface := range ifaces {
+		if strings.HasPrefix(iface.Name, prefix) {
+			logger.Warningf("cleaning up interface %q which still exists", iface.Name)
```

Comment:
How do we know if it's a leftover interface or an interface from another cluster? (future work). We might be removing something that is working.

Reply:
Definitely true! This is a complication I should note.